packages/core/src/modules/proofs/formats/indy/util.ts

```diff
@@ -0,0 +1,252 @@
+  nonce: string
+  attributes: V1PresentationPreviewAttributeOptions[]
+  predicates: V1PresentationPreviewPredicateOptions[]
+}): ProofRequest {
+  const proofRequest = new ProofRequest({
+    name,
+    version,
+    nonce,
+  })
+
+  /**
+   * Create mapping of attributes by referent. This required the
+   * attributes to come from the same credential.
+   * @see https://github.com/hyperledger/aries-rfcs/blob/master/features/0037-present-proof/README.md#referent
+   *
+   * {
+   *  "referent1": [Attribute1, Attribute2],
+   *  "referent2": [Attribute3]
+   * }
+   */
+  const attributesByReferent: Record<string, V1PresentationPreviewAttributeOptions[]> = {}
+  for (const proposedAttributes of attributes ?? []) {
+    if (!proposedAttributes.referent) proposedAttributes.referent = uuid()
+
+    const referentAttributes = attributesByReferent[proposedAttributes.referent]
+
+    // Referent key already exist, add to list
+    if (referentAttributes) {
+      referentAttributes.push(proposedAttributes)
+    }
+
+    // Referent key does not exist yet, create new entry
+    else {
+      attributesByReferent[proposedAttributes.referent] = [proposedAttributes]
+    }
+  }
+
+  // Transform attributes by referent to requested attributes
+  for (const [referent, proposedAttributes] of Object.entries(attributesByReferent)) {
+    // Either attributeName or attributeNames will be undefined
+    const attributeName = proposedAttributes.length == 1 ? proposedAttributes[0].name : undefined
+    const attributeNames = proposedAttributes.length > 1 ? proposedAttributes.map((a) => a.name) : undefined
+
+    const requestedAttribute = new ProofAttributeInfo({
+      name: attributeName,
+      names: attributeNames,
+      restrictions: [
+        {
+          credentialDefinitionId: proposedAttributes[0].credentialDefinitionId,
+        },
+      ],
+    })
+
+    proofRequest.requestedAttributes.set(referent, requestedAttribute)
+  }
+
+  // Transform proposed predicates to requested predicates
+  for (const proposedPredicate of predicates ?? []) {
+    const requestedPredicate = new ProofPredicateInfo({
+      name: proposedPredicate.name,
+      predicateType: proposedPredicate.predicate,
+      predicateValue: proposedPredicate.threshold,
+      restrictions: [
+        {
+          credentialDefinitionId: proposedPredicate.credentialDefinitionId,
+        },
+      ],
+    })
+
+    proofRequest.requestedPredicates.set(uuid(), requestedPredicate)
+  }
+
+  return proofRequest
+}
+
+/**
+ * Checks whether two `names` arrays are equal. The order of the names doesn't matter.
+ */
+function areNamesEqual(namesA: string[] | undefined, namesB: string[] | undefined) {
+  if (namesA === undefined) return namesB === undefined || namesB.length === 0
+  if (namesB === undefined) return namesA.length === 0
+
+  // Check if there are any duplicates
+  if (new Set(namesA).size !== namesA.length || new Set(namesB).size !== namesB.length) return false
```

Comment:
Hmm, this will actually make the check a lot more complex, as then we'd need to check the number of occurences for each property name and count how many we've already processed. So in this case it makes the implementation easier